src/assign.c

```diff
@@ -742,7 +742,7 @@ const char *memrecycle(const SEXP target, const SEXP where, const int start, con
           const double *sd = REAL(source);
           for (int i=0; i<slen; ++i) {
             const double val = sd[i+soff];
-            if (!ISNAN(val) && (!R_FINITE(val) || val!=(int)val || (int)val<1 || (int)val>nlevel)) {
+            if (!ISNAN(val) && (!within_int32_repres(val) || val!=(int)val || (int)val<1 || (int)val>nlevel)) {
```

Comment:
Do we want to avoid `(int)val` three times? Maybe

```c
const double val...
IF (... || ((int int_val=(int)val)!=val) || int_val<1 || int_val>n_level)
```

Or

```c
if (!ISNAN(val)) {
  if (within_int32_repres(val)) {
    error(...)
  } else {
    int_val = (int)val;
    if (val != int_val) {
      error(...)
    }
    if (int_val < 1 || int_val > nlevel) {
      error(...)
    }
  }
}
```

Don't feel strongly & my options are certainly a lot more verbose/code.

Reply:
I simplified the logic instead in a27ccf28

Reply:
FFR I think the commit is 9d05feb0f41a98c535568ad7ec1f328b45546916

Reply:
excellent, much simpler!